### PR TITLE
fixed invalid endpoint error

### PR DESCRIPTION
**Pull request by MonzElmasry**
the pr basically fixed two things:

- error out when passing dsn without " (driver) :// "
-  timeout with postgres case when it takes forever to do a db ping

And add drone automation 

https://github.com/rancher/k3s/issues/1668

### CODE REVIEW

**Comment by galal-hussein** [comment]
@MonzElmasry you might want to run `gofmt` against the whole code to fix the indentation and spaces for the code

**Comment by cjellick** [comment]
@MonzElmasry what happens if the url is technically correct but points to nothing, like the database doesnt actually exist or the network is down.

In this scenario, I was expecting k3s to actually fail to start, though I'm not even sure if that is the right behavior.  @ibuildthecloud should k3s fail to start or sit in an endless retry loop waiting for the db connection to be successful? Should we introduce some sort of connection test?

@MonzElmasry while we wait for @ibuildthecloud's response, can you investigate around making a test connection? Just some research and a prototype. I think there may be some established best practices around testing connections to sql dbs, but dont know for sure.

**Comment by cjellick** [comment]
@MonzElmasry I'd also like you to consider how we could potentially write tests for your changes. To my knowledge, kine doesnt currently have meaningful tests and I haven't put much thought into how we could introduce them. Maybe we pull the framework we have in k3s for hooking k3s up to containerized databases into this repository and start with some simple smoke tests of connecting to the db backends.

**Comment by davidnuzik** [comment]
So when we first talked about this GH issue I believe that yes we were suggesting that K3s should NOT start at all on the initial connection to the DB. If that connection fails, K3s should not run and a sensible error should be shown to the user.

File: Dockerfile.dapper [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+RUN mkdir -p /go/src/golang.org/x && \
+    cd /go/src/golang.org/x && git clone https://github.com/golang/tools && \
+    git -C /go/src/golang.org/x/tools checkout -b current aa82965741a9fecd12b026fbb3d3c6ed3231b8f8 && \
+    go install golang.org/x/tools/cmd/goimports
+RUN rm -rf /go/src /go/pkg
+RUN if [ "${ARCH}" == "amd64" ]; then \
+        curl -sL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s v1.15.0; \
+    fi
+
+RUN curl -L https://github.com/rancher/rke/releases/download/v1.0.5/rke_linux-$ARCH \
```
Comment (galal-hussein) [inline Dockerfile.dapper:20]: this tool is not needed in kine

File: Dockerfile.dapper [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    --output /usr/local/bin/rke && chmod +x /usr/local/bin/rke
+
+ENV GO111MODULE off
+ENV DAPPER_ENV REPO TAG DRONE_TAG
+ENV DAPPER_SOURCE /go/src/github.com/rancher/helm-controller/
```
Comment (galal-hussein) [inline Dockerfile.dapper:25]: this also need to be changed to rancher/kine

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -34,28 +35,58 @@ type Config struct {
...
+
+	type Response struct {
+		leaderelect bool
+		backend     server.Backend
+		err         error
+	}
+	ch := make(chan Response, 1)
+
+	go func() {
+		leaderelect, backend, err := getKineStorageBackend(ctx, driver, dsn, config)
```
Comment (briandowns) [inline pkg/endpoint/endpoint.go:66]:
```suggestion
		leaderElect, backend, err := getKineStorageBackend(ctx, driver, dsn, config)
```

File: pkg/endpoint/endpoint.go [hunk 4]
```diff
@@ -34,28 +35,58 @@ type Config struct {
...
+		res := Response{leaderelect: leaderelect, backend: backend, err: err}
+		ch <- res
+	}()
+
+	var (
+		leaderelect bool
```
Comment (briandowns) [inline pkg/endpoint/endpoint.go:72]:
```suggestion
		leaderElect bool
```

File: pkg/endpoint/endpoint.go [hunk 5]
```diff
@@ -34,28 +35,58 @@ type Config struct {
...
+	)
+
+	select {
+	case res := <-ch:
+		leaderelect = res.leaderelect
+		backend = res.backend
+		err = res.err
+
+	case <-time.After(10 * time.Second):
+		return ETCDConfig{}, fmt.Errorf("Timeout, please check your connection/ host port number")
```
Comment (briandowns) [inline pkg/endpoint/endpoint.go:83]:
```suggestion
		return ETCDConfig{},errors.New("Timeout, please check your connection/ host port number")
```

File: pkg/endpoint/endpoint.go [hunk 6]
```diff
@@ -34,28 +35,58 @@ type Config struct {
...
+	type Response struct {
+		leaderelect bool
+		backend     server.Backend
+		err         error
+	}
+	ch := make(chan Response, 1)
+
+	go func() {
+		leaderelect, backend, err := getKineStorageBackend(ctx, driver, dsn, config)
+		res := Response{leaderelect: leaderelect, backend: backend, err: err}
```
Comment (briandowns) [inline pkg/endpoint/endpoint.go:67]: This is a little more readable if split to multiple lines.

File: pkg/endpoint/endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+		{"postgres://123", "Timeout, please check your connection/ host port number"},
+		{"postgres://", "connection refused"},
+		{"text", "invalid endpoint value"},
+	}
+	for i := range endpoints {
+		config.Endpoint = endpoints[i][0]
+		_, err := Listen(ctx, config)
+		if err == nil {
+			t.Error("unpassed error !! ", endpoints[i][1])
+		} else {
```
Comment (briandowns) [inline pkg/endpoint/endpoint_test.go:32]: Is this else necessary?

File: pkg/endpoint/endpoint.go [hunk 8]
```diff
@@ -138,17 +167,21 @@ func getKineStorageBackend(ctx context.Context, driver, dsn string, cfg Config)
 	return leaderElect, backend, err
 }
 
-func ParseStorageEndpoint(storageEndpoint string) (string, string) {
+func ParseStorageEndpoint(storageEndpoint string) (string, string, error) {
 	network, address := networkAndAddress(storageEndpoint)
+	var err error = nil
```
Comment (briandowns) [inline pkg/endpoint/endpoint.go:172]:
```suggestion
	var err error
```

File: pkg/endpoint/endpoint_test.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	"context"
+	"fmt"
+	"testing"
+)
+
+func TestListen(t *testing.T) {
+	var config Config
+	var ctx context.Context
+	config.Listener = ""
+	endpoints := [13][2]string{{"postgres://doadmin:ax4t3h9twe04fa7k@postgressdb-do-user-898766-0.a.db.ondigitalocean.com:2560/defaultdb", "invalid port number"},
```
Comment (galal-hussein) [inline pkg/endpoint/endpoint_test.go:13]: This is not the right approcah for unit testing for this function, basically we want to make sure that the function returns an error when an invalid address is passed, so there are basically 3 cases:

- DSN with no <driver>:// and this should return the new error you implemented
- DSN with <driver>:// and invalid address and this should return an error
- DSN with correct <driver>:// and address

Comment (galal-hussein) [reply pkg/endpoint/endpoint_test.go:13]: you dont have to have a working mysql and postgres to test the last case, we basically have these integration tests in k3s https://github.com/rancher/k3s/blob/master/scripts/test-setup-sonobuoy-mysql

cc @cjellick

File: .golangci.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (cjellick) [inline .golangci.json:1]: I don think this file should be checked in, should it?

Comment (MonzElmasry) [reply .golangci.json:1]: it should be added 
https://github.com/rancher/k3s/blob/master/.golangci.json
because it's used in golint-ci in drone.

**Comment by davidnuzik** [comment]
@ibuildthecloud can we get final review and LGTM from you on this PR? cc: @cjellick

**Comment by ibuildthecloud** [comment]
Can this PR be split into two commits.  All the CI changes in one and the code changes in another.  Also, don't delete Dockerfile, that file is used.

File: pkg/endpoint/endpoint.go [hunk 11]
```diff
@@ -34,28 +35,60 @@ type Config struct {
...
 			Endpoints:   strings.Split(config.Endpoint, ","),
 			TLSConfig:   config.Config,
 			LeaderElect: true,
 		}, nil
 	}
+	if err != nil {
+		return ETCDConfig{}, errors.Wrap(err, "parsing storage point")
+	}
+
+	type Response struct {
```
Comment (ibuildthecloud) [inline pkg/endpoint/endpoint.go:57]: Sorry, I don't fully understand the context, but it really doesn't seem like the right place to actually be testing the connection.  What is the issue you are trying to fix that requires testing the connection?

File: pkg/endpoint/endpoint.go [hunk 12]
```diff
@@ -138,17 +169,21 @@ func getKineStorageBackend(ctx context.Context, driver, dsn string, cfg Config)
 	return leaderElect, backend, err
 }
 
-func ParseStorageEndpoint(storageEndpoint string) (string, string) {
+func ParseStorageEndpoint(storageEndpoint string) (string, string, error) {
 	network, address := networkAndAddress(storageEndpoint)
+	var err error
```
Comment (ibuildthecloud) [inline pkg/endpoint/endpoint.go:174]: Is it more straightfoward to just say
```go
if storageEndpoint == "" {
  return SQLiteBackend, ""
}
```
And change nothing else?

File: pkg/endpoint/endpoint_test.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"context"
+	"testing"
+)
+
+func TestListen(t *testing.T) {
+	var config Config
+	var ctx context.Context
+	config.Listener = ""
+	endpoints := [5]string{
```
Comment (ibuildthecloud) [inline pkg/endpoint/endpoint_test.go:12]:
```suggestion
	endpoints := []string{
```
The 5 is not needed.  Putting a number there creates an actual array and not a slice.  Arrays are using in go really only in very performance critical situations.

File: pkg/endpoint/endpoint_test.go [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package endpoint
+
+import (
+	"context"
+	"testing"
+)
+
+func TestListen(t *testing.T) {
+	var config Config
+	var ctx context.Context
```
Comment (ibuildthecloud) [inline pkg/endpoint/endpoint_test.go:10]:
```suggestion
	var (
		config Config
		ctx context.Context
	)
```
Just looks more idiomatic this way

File: pkg/endpoint/endpoint_test.go [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+	config.Listener = ""
```
Comment (ibuildthecloud) [inline pkg/endpoint/endpoint_test.go:11]: Is this line redundant?  Strings are initialized to "" by default.

**Pull request closed without merging**